### PR TITLE
HDDS-8153. Integrate ContainerBalancer with MoveManager

### PR DESCRIPTION
**Pull request by siddhantsangwan**
## What changes were proposed in this pull request?

Currently ContainerBalancer uses ReplicationManager#move which internally uses LegacyReplicationManager. This jira proposes integrating MoveManager with ContainerBalancer such that it uses MoveManager#move if "hdds.scm.replication.enable.legacy" is false and ReplicationManager#move if it's true.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8153

## How was this patch tested?

Added a UT and modified existing ones

### CODE REVIEW

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java [hunk 1]
```diff
@@ -757,48 +785,66 @@ private boolean moveContainer(DatanodeDetails source,
...
-
-            metrics.incrementCurrentIterationContainerMoveMetric(result, 1);
-            if (ex != null) {
-              LOG.info("Container move for container {} from source {} to " +
-                      "target {} failed with exceptions {}",
-                  containerID.toString(),
+
+      ReplicationManager.ReplicationManagerConfiguration rmConf =
+          ozoneConfiguration.getObject(
+              ReplicationManager.ReplicationManagerConfiguration.class);
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:791]: nit: I think this should be stored in the constructor.

**Review by adoroszlai — COMMENTED** [review]
LGTM so far.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java [hunk 2]
```diff
@@ -757,48 +785,66 @@ private boolean moveContainer(DatanodeDetails source,
...
+                  "target {} failed with exceptions.",
+              containerID.toString(),
+              source.getUuidString(),
+              moveSelection.getTargetNode().getUuidString(), ex);
+          metrics.incrementNumContainerMovesFailedInLatestIteration(1);
+        } else {
+          if (result == MoveManager.MoveResult.COMPLETED) {
+            sizeActuallyMovedInLatestIteration +=
+                containerInfo.getUsedBytes();
+            if (LOG.isDebugEnabled()) {
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerTask.java:817]: NIT: We usually don't need to wrap debug calls in `if (LOG.isDebugEnabled`) if the parameters to the log are simple getters, which I think they are here.

**Comment by siddhantsangwan** [comment]
Thanks for the reviews. I've addressed them in the latest commit.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java [hunk 3]
```diff
@@ -93,16 +93,16 @@ public final class ContainerBalancerConfiguration {
       "to exclude from balancing. For example \"1, 4, 5\" or \"1,4,5\".")
   private String excludeContainers = "";
 
-  @Config(key = "move.timeout", type = ConfigType.TIME, defaultValue = "30m",
+  @Config(key = "move.timeout", type = ConfigType.TIME, defaultValue = "60m",
```
Comment (siddhantsangwan) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:96]: Since the replication commands scheduled by `MoveManager` as part of `MoveManager#move` have lower priority compared to replication commands by `ReplicationManager`, I've increased this default to 60 minutes.

Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:96]: We can do this in another PR, but there is a timeout hard coded into MoveManager right now. Probably we need to pass this value into MoveManager somehow so it passes a sensible timeout to RM when scheduling the command. That will then set the DN deadline and the pending Ops timeout.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java [hunk 4]
```diff
@@ -93,16 +93,16 @@ public final class ContainerBalancerConfiguration {
...
       tags = {ConfigTag.BALANCER}, description =
       "The amount of time to allow a single container to move " +
           "from source to target.")
-  private long moveTimeout = Duration.ofMinutes(30).toMillis();
+  private long moveTimeout = Duration.ofMinutes(60).toMillis();
 
   @Config(key = "balancing.iteration.interval", type = ConfigType.TIME,
-      defaultValue = "70m", tags = {ConfigTag.BALANCER}, description =
+      defaultValue = "130m", tags = {ConfigTag.BALANCER}, description =
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java:103]: Should the timeout and interval not be quite close in time? If commands timeout after 60 mins, and the interval is 130m, does that mean the balancer will go idle for some time in between?

**Comment by siddhantsangwan** [comment]
I noticed a subtle bug in this PR. We're only constructing `MoveManager` once when `ContainerBalancerTask` is constructed. But `ContainerBalancerTask` works in iterations and resets any collections or members that hold state. Since `MoveManager` tracks moves using ContainerID as the key in a hash map, we need to reset its state between iterations too.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java (hunk 3 above)
Comment (siddhantsangwan) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancerConfiguration.java]: Yes, I've reverted any changes in this PR related to timeouts. We can do that in the next PR.

**Review by sodonnel — APPROVED** [review]
This changes are good to commit if we get green CI. 

I had a discussion with @siddhantsangwan and we know we need to take MoveManager construction out of ContainerBalancerTask and inject the dependency, and also make the MoveManager instance register with ContainerReplicaPendingOps.

We are going to do that in a followup PR immediately after this one.

**Comment by siddhantsangwan** [comment]
Thanks for the reviews. Merging now that CI is green. Follow up Jira: https://issues.apache.org/jira/browse/HDDS-8167
